# 4.4 IC Data Wrangling & Subsetting

### contents

### 1.0 Import libraries

### 2.0 Import Data

### 3.0 Wrangle Data

### 4.0 Task questions (wrangling some more data)

# 1.0 Import libraries

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 2.0 Import Data

## setup

In [2]:
#creates a shortcut to my project folder
path = r'C:\Users\Asus 17\OneDrive\Desktop\Achievement 4\instacart_basket_analysis'

## import orders

In [3]:
#import orders.csv as df_ords
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

## import products

In [4]:
#import products.csv as df_prods
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

## import departments


In [5]:
# import departments.csv as df_dep
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

# 3.0 Wrangle Data


## drop unnecessary columns


In [7]:
# drop eval_set from df_ords
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [8]:
# drop eval_set from df_ords and save it as df_ords_2 
df_ords_2 = df_ords.drop(columns = ['eval_set'])

In [9]:
# looking for missing values with basically a frequency table
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [10]:
# rename order_dow
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [11]:
# check it was done
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [12]:
# get some descriptive stats
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [13]:
# change order_id data type to a string
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [14]:
#check it happened
df_ords['order_id'].dtype

dtype('O')

## check out departments

In [15]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [16]:
# transpose df_dep
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [17]:
# save df_dep.T as df_dep_t
df_dep_t = df_dep.T

In [18]:
#check that happened
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [19]:
# add an index column
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

## Making a new header from the first row and dropping that 0


In [20]:
# naming the data from row 0
new_header = df_dep_t.iloc[0]

In [21]:
# checking it worked
new_header

0    department
Name: department_id, dtype: object

In [22]:
# taking everything from row 1 onwards and putting it in df_dep_t_new
df_dep_t_new = df_dep_t[1:]

In [23]:
#checking it worked
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [24]:
# setting new_header as the titles of the columns in my new df
df_dep_t_new.columns = new_header

In [25]:
#checking it worked
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

## turning df_dep_t_new into a Data Dictionary

In [26]:
#make a data dictionary
data_dict = df_dep_t_new.to_dict('index')

In [27]:
#checking it worked
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

## putting my dictionary to use

In [28]:
# show the top 5 rows of df_prods
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [29]:
# find out which department Chocolate Sandwich Cookies are in
print(data_dict.get('19'))

{'department': 'snacks'}


In [30]:
#checking for snacks True/False
df_prods['department_id'] == 19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [31]:
#looking just for snacks
df_prods[df_prods['department_id'] == 19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [32]:
# saving the snacks 
df_snacks = df_prods[df_prods['department_id'] == 19]

In [33]:
#checking it worked
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

## looking for snacks with loc

In [34]:
#making a new df for snacks using loc
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19
                          ]

In [35]:
#doing it again but with isin so we could find multiple departments if we wanted
#maybe snacks and seafood
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

# 4.4 Task 4.4 questions (wrangling some more data)

In [36]:
#check head of df_ords
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0  2539329        1    prior             1                   2   
1  2398795        1    prior             2                   3   
2   473747        1    prior             3                   3   
3  2254736        1    prior             4                   4   
4   431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [37]:
#check head of df_ords_2
df_ords_2.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [38]:
# drop eval_set from df_ords
df_ords = df_ords.drop(columns = ['eval_set'])

## question 2


In [39]:
#check df_ords data types
df_ords.dtypes

order_id                   object
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [40]:
# change user_id data type to string
df_ords['user_id'] = df_ords['order_id'].astype('str')

In [41]:
#check it worked
df_ords.dtypes

order_id                   object
user_id                    object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [42]:
#check singular for good measure
df_ords['order_id'].dtype

dtype('O')

## question 3

In [43]:
#rename 'orders_dow' to be a bit more intuitive
df_ords_2.rename(columns = {'order_dow' : 'order_day_of_week'}, inplace = True)

In [44]:
#check it worked
df_ords_2.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

## question 4


In [45]:
#check how often each hour of the day shows up, even NA
df_ords['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

10AM is the most popular hour of for orders with 288418 orders.



11AM, 3PM and 2PM are all also popular. Each with over 280000 orders. 


# question 5

In [46]:
# look up department 4 in the data dictionary I created above
print(data_dict.get('4'))

{'department': 'produce'}


A department id of '4' refers to the 'produce' department

# question 6

In [47]:
# check the  department number for 'breakfast'
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

Its '14'

In [48]:
#create a subset of df_prods containing only breakfast foods (department_id = '14')
df_breakfast = df_prods.loc[df_prods ['department_id'].isin([14])]

In [49]:
# check it worked
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

# question 7

 Dinner Party Department ids: alcohol = '5', deli = '20', beverages = '7', meat/seafood = '12'


In [50]:
#make a dataframe for the dinner party products
df_dinner_party = df_prods.loc[df_prods['department_id'].isin([5, 20, 7, 12])]

In [51]:
#check that it worked
df_dinner_party.head(20)

product_id                                       product_name  aisle_id  \
2            3               Robust Golden Unsweetened Oolong Tea        94   
6            7                     Pure Coconut Water With Orange        98   
9           10     Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                                  Peach Mango Juice        31   
16          17                                  Rendered Duck Fat        35   
19          20     Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
22          23                             Organic Turkey Burgers        49   
34          35     Italian Herb Porcini Mushrooms Chicken Sausage       106   
38          39           Daily Tangerine Citrus Flavored Beverage        64   
39          40  Beef Hot Links Beef Smoked Sausage With Chile ...       106   
43          44                        Sparkling Raspberry Seltzer       115   
48          49      Vegetarian Grain Meat Sausages Italian - 4 CT        14   
51          52                                Mirabelle Brut Rose       134   
52          53              Whole Leaf Pure Aloe With Lemon Juice        98   
61          62            Premium Deli Oven Roasted Turkey Breast        96   
68          69                                                NaN        26   
72          73                  Jasmine Tea Unfiltered Ginger Ale        77   
73          74                    Artisan Chick'n & Apple Sausage        14   
81          82                       All Natural 100% Apple Juice        98   
83          84                                         Lamb Shank         7   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1  
19              7     6.0  
22             12     8.2  
34             12    15.1  
38              7    12.5  
39             12    22.5  
43              7    12.2  
48             20    10.1  
51              5    14.4  
52              7     4.7  
61             20    14.6  
68              7    11.8  
72              7     8.3  
73             20     3.0  
81              7     6.4  
83             12    24.3

# question 8

In [128]:
# how many (rows, columns) does df_dinner_party have?
df_dinner_party.shape

(7650, 5)

It has 7650 rows


# Question 9


In [66]:
#make a df about user 1
df_user_1 = df_ords[df_ords['user_id'] == '1']

In [67]:
#check that it worked
df_user_1.head(10)

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
1868044        1       1             4                   4                 10   

         days_since_prior_order  
1868044                     9.0

In [56]:
#what is its shape?
df_user_1.shape

(1, 6)

In [59]:
# try to have another look at df_ords
df_ords[df_ords['user_id'] == '1'].head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
1868044        1       1             4                   4                 10   

         days_since_prior_order  
1868044                     9.0

In [68]:
# having a look by order number
df_ords.loc[df_ords['order_number'].isin([1,2,3,4,5,6,7,8,9,10])]

order_id  user_id  order_number  orders_day_of_week  \
0        2539329  2539329             1                   2   
1        2398795  2398795             2                   3   
2         473747   473747             3                   3   
3        2254736  2254736             4                   4   
4         431534   431534             5                   4   
...          ...      ...           ...                 ...   
3421074  3186442  3186442             6                   0   
3421075   550836   550836             7                   2   
3421076  2129269  2129269             8                   3   
3421077  2558525  2558525             9                   4   
3421078  2266710  2266710            10                   5   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421074                 16                     3.0  
3421075                 13                     9.0  
3421076                 17                    22.0  
3421077                 15                    22.0  
3421078                 18                    29.0  

[1680424 rows x 6 columns]

In [69]:
# having a look at df_ords_2 because obviously something isn't right
df_user_1_2 = df_ords_2.loc[df_ords_2['user_id'].isin([1])]

In [71]:
#checking its ok
df_user_1_2.head(20)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2398795        1             2                  3                  7   
2     473747        1             3                  3                 12   
3    2254736        1             4                  4                  7   
4     431534        1             5                  4                 15   
5    3367565        1             6                  2                  7   
6     550135        1             7                  1                  9   
7    3108588        1             8                  1                 14   
8    2295261        1             9                  1                 16   
9    2550362        1            10                  4                  8   
10   1187899        1            11                  4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

In [72]:
# describe df_user_1_2
df_user_1_2.describe()

order_id  user_id  order_number  order_day_of_week  \
count  1.100000e+01     11.0     11.000000          11.000000   
mean   1.923450e+06      1.0      6.000000           2.636364   
std    1.071950e+06      0.0      3.316625           1.286291   
min    4.315340e+05      1.0      1.000000           1.000000   
25%    8.690170e+05      1.0      3.500000           1.500000   
50%    2.295261e+06      1.0      6.000000           3.000000   
75%    2.544846e+06      1.0      8.500000           4.000000   
max    3.367565e+06      1.0     11.000000           4.000000   

       order_hour_of_day  days_since_prior_order  
count          11.000000               10.000000  
mean           10.090909               19.000000  
std             3.477198                9.030811  
min             7.000000                0.000000  
25%             7.500000               14.250000  
50%             8.000000               19.500000  
75%            13.000000               26.250000  
max            16.000000               30.000000

# question 10

Ok. Something has happened to df_ords that I can't quite figure out. It must have something to do with changing the datatype of that column?


Going by the information i have from df_ords_2:

User 1 has made 11 orders, They have an average of 19 days between orders, on average they order at 10AM.

# export dataframes

In [75]:
#export df_ords as orders_wrangled.csv
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [76]:
#export df_ords_2 as orders_2_wrangled.csv just in case
df_ords_2.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_2_wrangled.csv'))

In [78]:
# export df_dep_t_new as departments_wrangled.csv
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))